### PR TITLE
Article page fixes.

### LINKED ISSUE #7: Model for Comments
_From @Alonski on April 27, 2017 17:12_

Create model for comments
Make it a hasMany on articles? #13 

_Copied from original issue: Alonski/realworld-ember#19_

### LINKED ISSUE #14: Build Article Route + Page
Need to get the article Data and use it to dynamically show the page.
Currently is a static template.

### LINKED ISSUE #56: [Article page] Render markdown from server
### Features
* Render the markdown from the article as HTML
* Use `marked` npm library to render the markdown
* Sanitize the content and then use `htmlSafe`. See Octane version [here](https://github.com/patocallaghan/realworld-starter-kit/blob/4611cda2c1b9f391f976b111e2e23a833d3dd00c/app/models/article.js#L29-L32)

For better encapsulation we should add this behaviour to the article model.

Want to write tests?
* Model unit: marked computed property generates the correct HTML
* Integration: Markdown is rendered as HTML on the page correctly

image

### LINKED ISSUE #57: [Article page] Follow/unfollow profile
Similar to https://github.com/gothinkster/ember-realworld/issues/52

image

### LINKED ISSUE #58: [Article page] Favorite/unfavorite post
Similar to https://github.com/gothinkster/ember-realworld/issues/51

image

### LINKED ISSUE #59: [Article page] Post comment (only shown to logged in)
### Feature
* Comment field
* disable 'post comment' button when textarea is empty
* It is only shown to logged in users 

image

### API
Add Comments to an Article
POST /api/articles/:slug/comments

Example request body:
```
{
  "comment": {
    "body": "His name was my name too."
  }
}
```
Authentication required, returns the created Comment

Required field: body

### Want to write tests?
* Integration: button is disabled when no content
* Integration: Posting comment sends the correct HTTP request
* Integration: Comment form is only visible when logged in

### LINKED ISSUE #60: [Article page] Comments card at the bottom of the page
### Features
* Render comments at the bottom of the article
* Comments should load async, i.e. kick off a fetch on `init`

See similar implementation in the Octane version [here](https://github.com/patocallaghan/realworld-starter-kit/blob/master/app/templates/components/comments-section.hbs)

image

### LINKED ISSUE #63: [Article page] Delete comment button (only shown to comment's author)
### Features
* Button which deletes the comment
* Should only be shown if the current user is the comment author

image

Want to write tests?
* Integration: Button is only shown for comment authors
* Acceptance: Clicking the button deletes the article and redirects to the homepage

### PR DESCRIPTION
**Pull request by GCheung55**
Fixes #7 - Model for Comments

Fixes #14 - Build Article Route + Page.

Fixes #60 - [Article Page] Comments card at bottom of the article page.
- ~Comments are loaded in the article route's `afterModel` right now. I hope that's okay.~

Fixes #59 - [Article Page] Post comment (only shown to logged-in users) on article page.

Fixes #63 - [Article Page] Delete comment button only shows for comment's author.
- Added `comment` model and relationship to `article` model.

Fixes #58 - [Article Page] Favorite/unfavorite article.

Fixes #57 - [Article Page] Follow/unfollow article author.

Fixes #56 - [Article Page] Render markdown from server.
- Installed `ember-auto-import` to load `marked` lib. I chose to use`marked` because it was used in some of the other realworld apps.

The new generated files include test files but I have not written tests for them.

### CODE REVIEW

**Comment by GCheung55** [comment]
@Alonski I've added tests that you requested, however I did not add acceptance tests. If it's okay with you, I can add them in a different issue?

File: package.json [hunk 1]
```diff
@@ -32,6 +32,7 @@
     "@ember/optional-features": "^0.6.3",
     "babel-eslint": "^10.0.1",
     "broccoli-asset-rev": "^2.7.0",
+    "ember-auto-import": "^1.5.2",
```
Comment (GCheung55) [inline package.json]: Added `ember-auto-import` to easily import Node.js modules. In this case, it was to import `marked` for Markdown parsing.

File: package.json [hunk 2]
```diff
@@ -51,6 +52,8 @@
     "ember-qunit": "^4.4.1",
     "ember-resolver": "^5.0.1",
     "ember-route-action-helper": "^2.0.7",
+    "ember-sinon": "^4.0.0",
+    "ember-sinon-qunit": "^3.4.0",
```
Comment (GCheung55) [inline package.json]: Added `ember-sinon-qunit`, which automatically installed `ember-sinon` to help with stubbing methods in the tests.

File: package.json [hunk 3]
```diff
@@ -59,6 +62,7 @@
     "eslint-plugin-ember": "^6.3.0",
     "eslint-plugin-prettier": "3.0.1",
     "loader.js": "^4.7.0",
+    "marked": "^0.7.0",
```
Comment (GCheung55) [inline package.json]: Some of the other realworld projects were using `marked`.

Compared to `markdown-it` (34.3KB gzipped) and `showdown` (26.1KB gzipped), `marked` was the lightest weight, coming in at 8.8KB gzipped.

File: app/templates/article.hbs [hunk 4]
```diff
@@ -1,97 +1,92 @@
...
-        <button class="btn btn-sm btn-outline-primary">
-          <i class="ion-heart"></i> &nbsp; Favorite Post
-          <span class="counter">(29)</span>
-        </button>
-      </div>
+      <h1>{{model.title}}</h1>
+      {{article-meta
+        article=model 
+        onFavoriteArticle=(action 
+          (if session.isLoggedIn
```
Comment (GCheung55) [inline app/templates/article.hbs]: The `favorite` and `follow` buttons behave differently when the user is anonymous or logged in. I opted to do the logic check in the template as opposed to in the route actions themselves to keep the actions focused.

File: app/adapters/article.js [hunk 5]
```diff
@@ -0,0 +1,3 @@
+import ApplicationAdapter from './application';
```
Comment (Alonski) [inline app/adapters/article.js]: Is this file needed?
Isn't the default for adapters to inherit from the ApplicationAdapter?

File: app/adapters/application.js [hunk 6]
```diff
@@ -1,13 +1,27 @@
...
 const { errorsHashToArray, RESTAdapter } = DS;
 
 export default RESTAdapter.extend({
+  session: service(),
+
   host: config.API.host,
 
   namespace: 'api',
 
+  headers: computed('session.token', function() {
```
Comment (Alonski) [inline app/adapters/application.js]: Was this not working before? I remember we used simple auth for this, I think.

File: app/adapters/comment.js [hunk 7]
```diff
@@ -0,0 +1,22 @@
+import ApplicationAdapter from './application';
+
+export default ApplicationAdapter.extend({
+  pathForType() {
+    return 'articles';
+  },
+
+  urlForCreateRecord(modelName, snapshot) {
```
Comment (Alonski) [inline app/adapters/comment.js]: Does it make sense and is it possible to use `buildUrl` here for all three of these?
Something like:
```js
buildURL(type, id, record) {
        const namespace = get(this, "namespace");
        return this._super(type, id, record).replace(namespace, `${namespace}/${record.get("article.id")}/comments/${id}`);
    },
```

File: app/components/article-meta.js [hunk 8]
```diff
@@ -0,0 +1,15 @@
+import Component from '@ember/component';
+import { inject as service } from '@ember/service';
+
+export default Component.extend({
+  session: service(),
+  /**
+   * Method executed when "favorite" button is clicked.
+   */
+  onFavoriteArticle() {},
```
Comment (Alonski) [inline app/components/article-meta.js]: Is there a reason for these empty functions?

File: app/components/favorite-article.js [hunk 9]
```diff
@@ -13,7 +13,7 @@ export default Component.extend({
     if (this.session.isLoggedIn) {
       await this.article[operation]();
     } else {
-      this.router.transitionTo('sign-in');
+      this.router.transitionTo('login');
```
Comment (Alonski) [inline app/components/favorite-article.js]: Nice catch :)

File: app/helpers/transition-to.js [hunk 10]
```diff
@@ -0,0 +1,12 @@
+import Helper from '@ember/component/helper';
+import { inject as service } from '@ember/service';
+
+export default Helper.extend({
+  router: service(),
+
+  compute([routeName, ...params]) {
```
Comment (Alonski) [inline app/helpers/transition-to.js]: This is pretty cool. Still is this not overly complicated?
Is there a way to use {{link-to}} instead of this helper?

Not necessarily for or against just want to discuss.

File: app/routes/article.js [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+
+export default Route.extend({
+  session: service(),
+  store: service(),
+
+  model({ slug }) {
+    return this.store.findRecord('article', slug);
+  },
+
+  actions: {
```
Comment (Alonski) [inline app/routes/article.js]: I know that we have `route-actions` in other parts of the app. I initially used them because I thought they were cool :) 
Saying that, I think that we should remove them from the app.
Could you please move these actions to the controller?
It will also clean up the code a bit. No more `this.controller.set`

File: app/templates/article.hbs (hunk 4 above)
Comment (Alonski) [reply app/templates/article.hbs]: I wrote above but is there a way to not use the `transition-to` helper and instead to use a `link-to`? Maybe inside the `article-meta` component?

Comment (GCheung55) [reply app/templates/article.hbs]: The only way I can think of is to add the template logic to use a `link-to` helper if the user is logged out. So the `article-meta` component template would be like

```handlebars
{{#if isLoggedIn}}
    <button {{action "onFavoriteArticle"}}>Favorite Article</button>
{{else}}
    {{#link-to "login"}}Favorite Article{{/link-to}}
{{/if}}
```

I'm okay with doing this too.

File: app/adapters/article.js (hunk 5 above)
Comment (GCheung55) [reply app/adapters/article.js]: You're right. I tried deleting it and requests depending on the adapter are still working. I'll remove it.

File: app/adapters/application.js (hunk 6 above)
Comment (GCheung55) [reply app/adapters/application.js]: `ember-simple-auth` isn't being used in the app so I had to add it myself.

File: app/adapters/comment.js (hunk 7 above)
Comment (GCheung55) [reply app/adapters/comment.js]: Yes, there's a pattern to these "urlFor" methods that can be addressed with `buildURL`.

File: app/components/article-meta.js (hunk 8 above)
Comment (GCheung55) [reply app/components/article-meta.js]: I have these no-ops as placeholders. I can remove them if you'd like.

File: app/routes/article.js (hunk 11 above)
Comment (GCheung55) [reply app/routes/article.js]: I originally wrote them in the controller and moved it to the route when I noticed the route-action. I'll move them back.

File: app/templates/article.hbs (hunk 4 above)
Comment (GCheung55) [reply app/templates/article.hbs]: I've added this.

File: app/adapters/article.js (hunk 5 above)
Comment (GCheung55) [reply app/adapters/article.js]: Removed!

File: app/adapters/comment.js (hunk 7 above)
Comment (GCheung55) [reply app/adapters/comment.js]: Added the `buildURL` and removed the others.

File: app/helpers/transition-to.js (hunk 10 above)
Comment (GCheung55) [reply app/helpers/transition-to.js]: Since I've replaced usage of this with a `link-to`, I've removed the helper and related test file.

File: app/routes/article.js (hunk 11 above)
Comment (GCheung55) [reply app/routes/article.js]: Moved actions from route to controller, also removed `ember-route-actions-helper` since it's not used.

File: app/components/article-meta.js (hunk 8 above)
Comment (GCheung55) [reply app/components/article-meta.js]: I decided to just remove them and add code comments for them.
